README.md
fauxdoc
=======

[![Build Status](https://github.com/unt-libraries/fauxdoc/actions/workflows/do-checks-and-tests.yml/badge.svg?branch=main)](https://github.com/unt-libraries/fauxdoc/actions)

* [About](#about)
* [Installation](#installation)
* [Basic Usage](#basic-usage)
* [Contributing](#contributing)
* [License](#license)

## About

*`fauxdoc`* is designed to help you efficiently generate fake (faux) record or document (doc) data conforming to bespoke requirements.

Fauxdoc is tested with Python versions 3.7 and above, including 3.11. It has almost no external requirements: if you are using Python 3.7, it requires typing_extensions and importlib_metadata to provide features that were added in 3.8. Otherwise, it requires nothing but the standard library.

### Why not Faker or Mimesis?

[Faker](https://faker.readthedocs.io) and [Mimesis](https://mimesis.name) are established tools for generating fake data, and they are way more "batteries included" than Fauxdoc. Why not just implement a set of custom data providers for one of these?

#### Dynamically Generating Bespoke Data

Whereas other libraries make it dead simple to produce values that recognizably correspond to real-world items or properties (colors, names, addresses, etc.), Fauxdoc helps you dial in on patterns or features that may only pertain to your use case. This is helpful if you're trying to test something specific, like forcing certain sets of edge cases.

Fauxdoc began as part of a utility for helping test and benchmark configurations for particular Solr collections. We wanted to test search performance by producing text that shared certain features of a live collection, such as using specific alphabets; having word, phrase, and/or sentence lengths (etc.) within certain limits; and having specific terms occur in certain specific distributions. And we wanted to be able to simulate facets by choosing data values from a finite list of random terms that would produce a term distribution similar to the real data. Even with Faker or Mimesis, we would have had to build most of this from scratch, anyway.

#### Field / Schema / Document-Set Control

Other libraries mostly focus on generating values in isolation, but Fauxdoc facilitates having more control at the `Field` and `Schema` levels. For example, when generating documents to use to benchmark Solr, we found that we wanted to be able to do things like control uniqueness on a per-field basis, control uniqueness across an entire document set, generate data based on values in other fields, and have better control over multi-valued fields.

#### Performant Extensibility

We wrote Fauxdoc knowing that we'd be using it to generate hundreds of thousands or millions of fake Solr documents at one time, so performance was a concern. However, Fauxdoc is also meant to be highly extensible, and it's easy for extensibility to come at the expense of performance. So, Fauxdoc classes are designed to allow performant extensibility. You generally have a choice: you can implement something as (e.g.) a wrapper that's conceptually simple but a bit slower, or you can implement the same feature using a custom class, with lower-level methods that are faster but a little less simple. It just depends on your use case and where you need the extra performance.

The built-in data providers (called `Emitters`) are designed to be as fast as we could make them. Their performance is roughly comparable to Mimesis', although this is an apples-to-oranges comparison. Like Mimesis, they are much faster than Faker.

[Top](#top)


## Installation

Install the latest published version of fauxdoc with:

```
python -m pip install fauxdoc
```

See [Contributing](#contributing) for the recommended installation process if you want to develop on fauxdoc. 

[Top](#top)


## Basic Usage

### Emitters

Conceptually, `Emitters` are like Faker or Mimesis `Providers`. They are the objects that output your data values: simply instantiate one and then call it. If you need multiple values at once, you can supply an integer when calling.

```python
from fauxdoc import emitters
myrandom = emitters.Choice(['a', 'b', 'c'], weights=[45, 45, 10])

myrandom()
# 'b'

myrandom(10)
# ['a', 'b', 'a', 'c', 'b', 'b', 'a', 'a', 'a', 'b'] 
```

Several emitter types are provided in `fauxdoc.emitters` that have general behavior and options. Above, the `Choice` emitter chooses randomly between multiple values, with optional weights and parameters to control uniqueness.

For more complex behavior, you can of course create your own Emitter classes using `fauxdoc.emitter.Emitter` as your base class. Mixins are provided in `fauxdoc.mixins` for standard ways of doing things (such as randomization).

### Fields

Each `Field` wraps an emitter instance and provides options to gate the output and/or generate multiple values. These options are themselves implemented as emitters. As with emitter instances, you also call a field instance to output values.

```python
from fauxdoc import emitters, profile

user_tags = ['adventure', 'yellow', 'awesome', 'food', 'action films']
user_tags_field = profile.Field(
  'user_tags',
  emitters.Choice(user_tags, replace_only_after_call=True),
  gate=emitters.chance(0.8),
  repeat=emitters.poisson_choice(range(1, 6), mu=3)
)

user_tags_field()
# ['action films', 'food', 'yellow']

user_tags_field()
# ['adventure', 'yellow', 'awesome', 'food']

user_tags_field()
# ['yellow', 'awesome', 'food']

user_tags_field()
# ['food', 'adventure', 'awesome', 'yellow']

user_tags_field()
# 
```

### Schema

Your `Schema` is a specific collection of field instances. Calling the schema instance generates data representing one full document (returned as a dictionary).

```python
import itertools
from fauxdoc import emitters, profile, dtrange

ENGLISH = emitters.make_alphabet([(ord('a'), ord('z'))])
GENRES = ['Science', 'Literature', 'Medicine', 'Fiction', 'Television']

myschema = profile.Schema(
  profile.Field('id', emitters.Iterative(lambda: itertools.count(1))),
  profile.Field(
    'title',
    emitters.WrapOne(
      emitters.Text(
        numwords_emitter=emitters.poisson_choice(range(1, 10), mu=2),
        word_emitter=emitters.Word(
          length_emitter=emitters.poisson_choice(range(1, 10), mu=5),
          alphabet_emitter=emitters.Choice(ENGLISH)
        )
      ),
      lambda title: title.capitalize()
    )
  ),
  profile.Field('doc_type', emitters.Choice(['report', 'article', 'book'])),
  profile.Field('date_created', emitters.Choice(dtrange.dtrange('1950-01-01', '2025-01-01'))),
  profile.Field(
    'genres',
    emitters.Choice(GENRES, replace_only_after_call=True),
    gate=emitters.chance(0.5),
    repeat=emitters.poisson_choice(range(1, 3), mu=1)
  )
)

myschema()
# {
#   'id': 1,
#   'title': 'Dvcoqh zbuaba',
#   'doc_type': 'book',
#   'date_created': datetime.date(1951, 8, 15),
#   'genres': ['Medicine', 'Fiction']
# }

myschema()
# {
#   'id': 2,
#   'title': 'Dird',
#   'doc_type': 'report',
#   'date_created': datetime.date(1998, 4, 6),
#   'genres': ['Fiction']
# }

myschema()
# {
#   'id': 3,
#   'title': 'Wvlptqk',
#   'doc_type': 'book',
#   'date_created': datetime.date(1977, 12, 10),
#   'genres': None
# }

myschema()
# {
#   'id': 4,
#   'title': 'Tnhkez',
#   'doc_type': 'article',
#   'date_created': datetime.date(1988, 1, 22),
#   'genres': None
# }

myschema()
# {
#   'id': 5,
#   'title': 'Ld gudv lnaxx',
#   'doc_type': 'article',
#   'date_created': datetime.date(1989, 9, 30),
#   'genres': ['Medicine']
# }
```

### Generating Data Based on Other Fields

For complex schemas, you may find generating values for each field in isolation to be too limiting. Fauxdoc allows you to create emitters that can access values generated in other fields. You can also create hidden fields, allowing you to generate a normalized or collective data value and then pull it into the appropriate de-normalized fields.

```python
import itertools
from fauxdoc import emitter, emitters, profile

def item_data_generator():
  for num in itertools.count(1):
    yield {
      'item_id': num,
      'barcode': 2000000000 + num
    }

myschema = profile.Schema(
  # This field is hidden. It generates data for 1 to 10 "items" that
  # the other fields then pull from.
  profile.Field(
    '__all_items',
    emitters.Iterative(item_data_generator),
    repeat=emitters.poisson_choice(range(1, 10), mu=3),
    hide=True,
  )
)

myschema.add_fields(
  profile.Field(
    'display_items',
    emitters.BasedOnFields(
      myschema.fields['__all_items'],
      lambda items: items[:3]
    )
  ),
  profile.Field(
    'more_items',
    emitters.BasedOnFields(
      myschema.fields['__all_items'],
      lambda items: items[3:] if len(items) > 3 else None
    )
  ),
  profile.Field(
    'has_more_items',
    emitters.BasedOnFields(
      myschema.fields['__all_items'],
      lambda items: bool(len(items) > 3)
    )
  ),
  profile.Field(
    'item_ids',
    emitters.BasedOnFields(
      myschema.fields['__all_items'],
      lambda items: [i['item_id'] for i in items]
    )
  ),
  profile.Field(
    'item_barcodes',
    emitters.BasedOnFields(
      myschema.fields['__all_items'],
      lambda items: [i['barcode'] for i in items]
    )
  )
)

myschema()
# {
#   'display_items': [
#     {'item_id': 1, 'barcode': 2000000001}
#   ],
#   'more_items': None,
#   'has_more_items': False,
#   'item_ids': [1],
#   'item_barcodes': [2000000001]
# }

myschema()
# {
#   'display_items': [
#     {'item_id': 2, 'barcode': 2000000002},
#     {'item_id': 3, 'barcode': 2000000003},
#     {'item_id': 4, 'barcode': 2000000004}
#   ],
#   'more_items': [
#     {'item_id': 5, 'barcode': 2000000005}
#   ],
#   'has_more_items': True,
#   'item_ids': [2, 3, 4, 5],
#   'item_barcodes': [2000000002, 2000000003, 2000000004, 2000000005]
# }
```

[Top](#top)


## Contributing

### Installing for Development and Testing

Fork the project on GitHub and then clone it locally:

```bash
git clone https://github.com/[your-github-account]/fauxdoc.git
```

All dependency and build information is defined in `pyproject.toml` and follows [PEP 621](https://peps.python.org/pep-0621/). From the fauxdoc root directory, you can install it as an editable project into your development environment with:

```bash
python -m pip install -e .[dev]
```

(The `[dev]` ensures it includes the optional development dependencies, namely pytest.)


### Running Tests

Run the full test suite in your active environment by invoking:

```bash
pytest
```

from the fauxdoc root directory.


#### Tox

Because this is a library, it needs to be tested against all supported environments for each update, not just one development environment. The tool we use for this is [tox](https://tox.wiki/en/latest/).

Rather than use a separate `tox.ini` file, I've opted to put the tox configuration directly in `pyproject.toml` (under the `[tool.tox]` table). There, I've defined several environments: flake8, pylint, and each of py37 through py311 using both the oldest possible dependencies and newest possible dependencies. When you run tox, you can target a specific environment, a specific list of environments, or all of them.

When tox runs, it automatically builds each virtual environment it needs, and then it runs whatever commands it needs within that environment (for linting, or testing, etc.). All you have to do is expose all the necessary Python binaries on the path, and tox will pick the correct one. My preferred way to manage this is with [pyenv](https://github.com/pyenv/pyenv) + [pyenv-virtualenv](https://github.com/pyenv/pyenv-virtualenv).

For example: Install these tools along with the Python versions you want to test against. Then:

1. Create an environment with tox installed. E.g.:
    ```
    pyenv virtualenv 3.10.8 tox-3.10.8
    pyenv activate
    python -m pip install tox
    ```
2. In the fauxdoc project repository root, create a file called `.python-version`. Add all of the Python versions you want to use, e.g., 3.7 to 3.11. For 3.10, use your `tox-3.10.8`. This should look something like this:
    ```
    3.7.15
    3.8.15
    3.9.15
    tox-3.10.8
    3.11.0
    ```
4. If `tox-3.10.8` is still activated, issue a `pyenv deactivate` command so that pyenv picks up what's in the file. (A manually-activated environment overrides anything set in a `.python-version` file.)
5. At this point you should have all five environments active at once in that directory. When you run `tox`, the tox in your `tox-3.10.8` environment will run, and it will pick up the appropriate binaries automatically (`python3.7` through `python3.11`) since they're all on the path.

Now you can just invoke tox to run linters and all the tests against all the environments:

```bash
tox
```

Or just run linters:

```bash
tox -e flake8,pylint_critical
```

Or run tests against a list of specific environments:

```bash
tox -e py39-oldest,py39-newest
```

[Top](#top)


## License

See the [LICENSE](LICENSE) file.

[Top](#top)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [v1.1.0](https://github.com/unt-libraries/fauxdoc/compare/v1.0.0...v1.1.0) — 2023-02-27

Overview of what's new in this version:
- `fauxdoc` is now fully typed and passes `mypy --strict` checks.
- Much refactoring. Fixing type hints revealed some underlying issues, and I've applied fixes and changes while leaving the public API as intact as possible — see details below about what exactly has changed.
- A small number of deprecations to be removed in v2. See details below.
- Previously undefined behavior, mainly around instance attributes, has been defined, documented, and tested. Attributes that are not read-only now have defined behavior when they are set after object instantiation.

### Added

- A `py.typed` file to indicate that the package now has type hints.
- New `mypy_strict` tox environment for running `mypy` tests.
- New tox environments for testing a package built using `build_package` against a target Python version: `py37-test_built_package` through `py311-test_built_package`.
- A new kwarg for initializing `fauxdoc.emitters.choice.Choice` instances: `cum_weights` can now be supplied, if needed. Note that you can only supply `weights` or `cum_weights` — not both. All previous args/kwargs still work as before.
- A `fauxdoc.emitters.fromfields.SourceFieldGroup` class. This is a subtype of `fauxdoc.groups.ObjectGroup`, used in `fauxdoc.emitters.fromfield` emitters to represent groups of source fields. It now implements the `single_valued` property that was previously only a private attribute on the `fauxdoc.emitters.fromfields.CopyFields` class.
- Additions to `fauxdoc.emitters.fromfields` emitters (`CopyFields` and `BasedOnFields`):
    - A new **public** attribute, `single_valued`, for both `CopyFields` and `BasedOnFields`. This attribute was previously private.
    - `set_source_fields` method, on `CopyFields` and `BasedOnFields` — used to set the `source` attribute from one or a list of `Field` instances. (You can set `source` directly, but that requires a `fauxdoc.emitters.fromfields.SourceFieldGroup` instance. This is just for convenience.)
    - `set_action_function` method, on `BasedOnFields` — used to set the `action` attribute from a function. (You can set `action` directly, but that requires a `fauxdoc.emitters.wrappers.BoundWrapper` instance. This is just for convenience.)
- Two new static methods on `fauxdoc.emitters.fixed.Iterative`, which move previously internal operations into utilities that can be used by external agents:
    - `check_iter_factory` — validates that an iterator factory function/method does not return an empty iterator, and raises a ValueError if it does.
    - `make_infinite_iter` — creates an iterator from an iterator factory that loops infinitely.
- A new `check_seq_iter_factory` static method on `fauxdoc.emitters.fixed.Sequential` allows validating that an iterator factory function/method probably generates an iterator that iterates over a sequence.
- A `fauxdoc.emitters.wrappers.BoundWrapper` class. Used to encapsulate a user-provided wrapper function, bind it to an object that provides RNG, and allow validating a given signature against the wrapper function's signature.
- A `set_wrapper_function` method on `fauxdoc.emitters.wrappers.WrapOne` and `fauxdoc.emitters.wrappers.WrapMany` — used to set the `wrapper` attribute from a function or other callable. (You can set `wrapper` directly, but that requires a `fauxdoc.emitters.wrappers.BoundWrapper` instance. This is just for convenience.)
- A new `set_fields` method, on `fauxdoc.profile.Schema` — used to set `fields` more naturally, without forcing the user to supply a `fauxdoc.groups.ObjectMap` instance. (You can still set `fields` directly, as an `ObjectMap`.)
- A `fauxdoc.typing.ImplementsRNG` protocol, for representing types that implement an `rng` attribute and `seed` method — such as `Field`s and various emitters.
- New `fauxdoc.typing.OrderedDict` and `fauxdoc.typing.UserList` types, to control for the fact that the `collections` versions of these types are not subscriptable in Python 3.7 and 3.8 (which is needed for type hints). For Python 3.7 and 3.8, a `__getitem__` method is monkey-patched onto `collections.OrderedDict` and `collections.UserList`, to make them subscriptable.
- New type aliases in `fauxdoc.typing`:
    - `F` — a TypeVar for float types.
    - `CT` — general-purpose covariant TypeVar.
    - `SourceT` — a contravariant TypeVar for e.g. callable arguments and other types that represent source data.
    - `OutputT` — a covariant TypeVar for e.g. callable return values that represent output data.
    - `FieldReturn` — represents the return type when calling `fauxdoc.profile.Field` instances.
- A new `fauxdoc.warn` module, which contains a function (`get_deprecated_attr`) for injecting a deprecation warning when a user gets a deprecated module or object attribute. (It's called from the `__getattr__` method of the applicable module or object.)

### Changed

- The minimum `pytest` version for Python 3.7 through 3.9 has been bumped up from 3.0.0 to 3.8.0. This is to allow testing deprecation warnings.
- Tox environment `build_package` no longer tries `pip install`ing the built package. It's assumed you'll use the new `*-test_built_package` environments to test the built package, which will of course have to install it.
- Comparing a `fauxdoc.dtrange.DateOrTimeRange` instance to another type of object is explicitly not supported and returns `NotImplemented`.
- On `fauxdoc.emitters.choice.Choice`, any sequences provided for the `weights` and `cum_weights` attributes are now cast to tuples on assignment so that they are immutable once assigned. (You can set a brand new sequence, but you cannot change individual elements.)

src/fauxdoc/typing.py
"""Contains constants/variables/etc. used for type hinting."""
import collections
import random
import sys
from typing import (
    Any, Callable, List, Optional, overload, Sequence, TYPE_CHECKING, TypeVar,
    Union
)

from fauxdoc.warn import get_deprecated_attr

if sys.version_info >= (3, 8):
    from typing import Protocol
else:
    from typing_extensions import Protocol

# For Python 3.7 and 3.8: standard library ABC classes (including
# collections) aren't subscriptable and don't accept e.g. UserList[T]
# at runtime, although they're required for type hints. This is the
# fix.
#
# On Python >= 3.9, or during type checking, the collections versions
# of these classes work.
if sys.version_info >= (3, 9) or TYPE_CHECKING:
    OrderedDict = collections.OrderedDict
    UserList = collections.UserList

# Otherwise, we need to hack it so these classes are subscriptable but
# are otherwise identical.
else:
    class _OrderedDict:
        def __getitem__(self, *args):
            return collections.OrderedDict

    class _UserList:
        def __getitem__(self, *args):
            return collections.UserList

    OrderedDict = _OrderedDict()
    UserList = _UserList()


# Type aliases defined here.

T = TypeVar('T')
F = TypeVar('F', bound=float)
CT = TypeVar('CT', covariant=True)
SourceT = TypeVar('SourceT', contravariant=True)
OutputT = TypeVar('OutputT', covariant=True)
FieldReturn = Optional[Union[T, List[T]]]


# Protocols defined here.

class EmitterLike(Protocol[T]):
    """Is like an emitter.Emitter object."""

    @property
    def num_unique_values(self) -> Optional[int]:
        ...

    @property
    def emits_unique_values(self) -> bool:
        ...

    @overload
    def __call__(self, number: None = None) -> T:
        ...

    @overload
    def __call__(self, number: int) -> List[T]:
        ...

    def __call__(self, number: Optional[int] = None) -> Union[T, List[T]]:
        ...

    def reset(self) -> None:
        ...


class ImplementsRNG(Protocol):
    """Is a type that implements RNG."""

    rng: random.Random

    def seed(self, rng_seed: Any) -> None:
        ...


class RandomEmitterLike(ImplementsRNG, EmitterLike[T], Protocol[T]):
    """Is like an emitter.RandomEmitter object."""


class FieldLike(Protocol[CT]):
    """Is like a profile.Field object."""

    multi_valued: bool
    name: str
    hide: bool

    def __call__(self) -> FieldReturn[CT]:
        ...

    def reset(self) -> None:
        ...

    @property
    def previous(self) -> FieldReturn[CT]:
        ...


# Handle deprecated module attributes.
# (Note that this is placed at the end of the module so we can use vars
# defined earlier in the module, in DEPRECATED.)

_deprecated_EmitterLike = Union[Callable[[int], Sequence[T]], EmitterLike[T]]

DEPRECATED = {
    'Number': ('float', float),
    'EmitterLikeCallable': ('EmitterLike', _deprecated_EmitterLike),
    'StrEmitterLike': ('EmitterLike[str]', EmitterLike[str]),
    'IntEmitterLike': ('EmitterLike[int]', EmitterLike[int]),
    'BoolEmitterLike': ('EmitterLike[bool]', EmitterLike[bool])
}


def __getattr__(name: str) -> Any:
    return get_deprecated_attr(name, __name__, 'module', DEPRECATED)


def __dir__() -> List[str]:
    return sorted(list(globals()) + list(DEPRECATED.keys()))
